packages/flutter_driver/lib/src/driver/timeline_summary.dart

```diff
@@ -458,4 +505,73 @@ class TimelineSummary {
   RasterCacheSummarizer _rasterCacheSummarizer() => RasterCacheSummarizer(_extractNamedEvents(kRasterCacheEvent));
 
   GCSummarizer _gcSummarizer() => GCSummarizer.fromEvents(_extractEventsWithNames(kGCRootEvents));
+
+  List<Duration> _extractOverallDurations() {
+    final List<_DurationTuple> buildDurationTuples =
+        _extractEventPairs(kBuildFrameEventName, _DurationTuple.from);
+    final List<_DurationTuple> rasterDurationTuples =
+        _extractEventPairs(kRasterizeFrameEventName, _DurationTuple.from);
+    buildDurationTuples.checkSortedByFirst();
+    rasterDurationTuples.checkSortedByFirst();
+
+    final List<_DurationTuple> pipelineItemTuples =
+        _extractNamedEvents(_kPipelineItemFrameEventName)
+            .groupListsBy((TimelineEvent e) => e.json['id'] as String)
+            .values
+            .map((List<TimelineEvent> events) => _DurationTuple.maybeFrom(
+                events.firstWhereOrNull((TimelineEvent e) => e.phase == 't'),
+                events.firstWhereOrNull((TimelineEvent e) => e.phase == 'f')))
+            .whereNotNull()
+            .sortedBy((_DurationTuple tuple) => tuple.first);
+
+    final List<Duration> ans = <Duration>[];
+    for (final _DurationTuple pipelineItemTuple in pipelineItemTuples) {
+      final _DurationTuple? buildDurationTuple =
+          buildDurationTuples.findItemThatContains(pipelineItemTuple.first);
+      final _DurationTuple? rasterDurationTuple =
+          rasterDurationTuples.findItemThatContains(pipelineItemTuple.second);
+      if (buildDurationTuple == null || rasterDurationTuple == null) {
+        continue;
+      }
+      ans.add(rasterDurationTuple.second - buildDurationTuple.first);
+    }
+
+    return ans;
+  }
+}
+
+class _DurationTuple {
+
+  const _DurationTuple(this.first, this.second);
+
+  factory _DurationTuple.from(TimelineEvent first, TimelineEvent second) =>
+      _DurationTuple(Duration(microseconds: first.timestampMicros!), Duration(microseconds: second.timestampMicros!));
+
+  final Duration first, second;
+
+  static _DurationTuple? maybeFrom(TimelineEvent? first, TimelineEvent? second) =>
+      (first == null || second == null) ? null : _DurationTuple.from(first, second);
+
+  bool contains(Duration d) => first <= d && d <= second;
+
+  @override
+  String toString() => 'DurationTuple($first, $second)';
+}
+
+extension on List<_DurationTuple> {
```

Comment:
Style guide says to not use extensions.

Also, why have a special tuple class at all instead of using records/tuples?

Reply:
Sure! Again this PR is ancient ;) Before the existence of records